src/paint.js

```diff
@@ -102,13 +102,6 @@ export default function setupPaint({
             );
           } catch (error) {
             console.log(error);
-          }
-          finally{
-            // Sense's undo/redo function waits for this promise to resolve,
-            // if it doesn't resolve, you cannot undo.
-            // There does not seem to be a reason for the code above to be inside a promise,
-            // and it does not always resolve after repainting.
-            // Therefore, we have resorted to the cowboy hack below. Enjoy!
             resolve();
```

Comment:
hmm, this is strange. Does it really work?
Before: resolve() runs in finally, so should always run, regardless or errors or not.
Now: resolve() runs in the catch, so only runs if there is an error?

Reply:
The problem is that like you said resolve() always runs. in paint.js directly. So, object.js in the client returns the onRendered callback too early before the object is done rendering. There is some code in PrintResolver to statisticBlock.js that is resolved when it is actually done. But this is too late if we already have called resolve() in the finally{} block.